File: backend/src/main/java/com/trio/backend/service/MessageService.java
package com.trio.backend.service;

import com.trio.backend.dto.communication.CreateMessageRequest;
import com.trio.backend.dto.communication.MessageResponse;
import com.trio.backend.dto.communication.UpdateMessageRequest;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.UUID;

public interface MessageService {

    MessageResponse create(UUID workspaceId, UUID conversationId, CreateMessageRequest request);

    MessageResponse getById(UUID workspaceId, UUID messageId);

    Page<MessageResponse> listByConversation(UUID workspaceId, UUID conversationId, UUID cursor, Pageable pageable);

    Page<MessageResponse> listBySender(UUID workspaceId, Pageable pageable);

    List<MessageResponse> listPinned(UUID workspaceId, UUID conversationId);

    Page<MessageResponse> listFilesByConversation(UUID workspaceId, UUID conversationId, Pageable pageable);

    Page<MessageResponse> listFilesByWorkspace(UUID workspaceId, Pageable pageable);

    Page<MessageResponse> search(UUID workspaceId, UUID conversationId, String query, Pageable pageable);

    Page<MessageResponse> searchByWorkspace(UUID workspaceId, String query, Pageable pageable);

    MessageResponse update(UUID workspaceId, UUID messageId, UpdateMessageRequest request);

    void delete(UUID workspaceId, UUID messageId);
}

File: frontend/project/src/services/message-service.ts
import { apiClient } from '../lib/api';
import type { PageResponse } from '../types/api';
import type {
  MessageResponse,
  CreateMessageRequest,
  UpdateMessageRequest,
} from '../types/communication';

function base(wsId: string, convId: string) {
  return `/workspaces/${wsId}/conversations/${convId}/messages`;
}

export const messageService = {
  list: (wsId: string, convId: string, cursor?: string) => {
    const params = cursor ? `?cursor=${cursor}` : '';
    return apiClient.get<PageResponse<MessageResponse>>(`${base(wsId, convId)}${params}`);
  },

  getById: (wsId: string, msgId: string) =>
    apiClient.get<MessageResponse>(`/workspaces/${wsId}/messages/${msgId}`),

  create: (wsId: string, convId: string, data: CreateMessageRequest) =>
    apiClient.post<MessageResponse>(`${base(wsId, convId)}`, data),

  update: (wsId: string, convId: string, msgId: string, data: UpdateMessageRequest) =>
    apiClient.put<MessageResponse>(`${base(wsId, convId)}/${msgId}`, data),

  delete: (wsId: string, convId: string, msgId: string) =>
    apiClient.delete<void>(`${base(wsId, convId)}/${msgId}`),

  listPinned: (wsId: string, convId: string) =>
    apiClient.get<MessageResponse[]>(`${base(wsId, convId)}/pinned`),

  listFiles: (wsId: string, convId: string) =>
    apiClient.get<PageResponse<MessageResponse>>(`${base(wsId, convId)}/files`),

  search: (wsId: string, convId: string, query: string) =>
    apiClient.get<PageResponse<MessageResponse>>(`${base(wsId, convId)}/search?query=${encodeURIComponent(query)}`),
};
